File: old/Assembly-CSharp_V1/SRPG/UnitParam.cs
// Decompiled with JetBrains decompiler
// Type: SRPG.UnitParam
// Assembly: Assembly-CSharp, Version=1.2.0.0, Culture=neutral, PublicKeyToken=null
// MVID: 9BA76916-D0BD-4DB6-A90B-FE0BCC53E511
// Assembly location: C:\Users\André\Desktop\Assembly-CSharp.dll

using GR;
using System;

namespace SRPG
{
  public class UnitParam
  {
    private static EElement[] WeakElements = new EElement[7]{ EElement.None, EElement.Water, EElement.Thunder, EElement.Fire, EElement.Wind, EElement.Dark, EElement.Shine };
    private static EElement[] ResistElements = new EElement[7]{ EElement.None, EElement.Wind, EElement.Fire, EElement.Thunder, EElement.Water, EElement.None, EElement.None };
    public OString ai = (OString) ((string) null);
    public OString birth = (OString) ((string) null);
    public OString grow = (OString) ((string) null);
    public OString piece = (OString) ((string) null);
    public OString skill = (OString) ((string) null);
    public OString ability = (OString) ((string) null);
    public OString ma_quest = (OString) ((string) null);
    public OInt sw = (OInt) 1;
    public OInt sh = (OInt) 1;
    public OInt rare = (OInt) 0;
    public OInt raremax = (OInt) 0;
    public OInt hero = (OInt) 0;
    public OInt search = (OInt) 0;
    public OBool stopped = (OBool) false;
    public DateTime available_at = DateTime.MinValue;
    public string[] leader_skills = new string[RarityParam.MAX_RARITY];
    public string[] recipes = new string[RarityParam.MAX_RARITY];
    private string localizedNameID;
    private string localizedTagID;
    private string tag;
    public int no;
    public string iname;
    public string name;
    public string model;
    public OString[] jobsets;
    public string[] tags;
    public ESex sex;
    public EUnitType type;
    public EElement element;
    public OInt height;
    public OInt weight;
    public bool summon;
    public string image;
    public string voice;
    public BaseStatus ini_status;
    public BaseStatus max_status;
    public OString default_skill;
    public OString[] default_abilities;
    public OString djob;
    public OString dbuki;
    public JobTypes jobtype;
    public RoleTypes role;
    public OInt mov;
    public OInt jmp;
    public OInt inimp;
    public OInt ma_rarity;
    public OInt ma_lv;
    public OString[] skins;
    public OString[] job_option_index;
    public OString[] job_images;
    public OString[] job_voices;
    public OBool is_throw;
    private JobSetParam[] mJobSetCache;

    protected void localizeFields(string language)
    {
      this.init();
      this.name = LocalizedText.SGGet(language, GameUtility.LocalizedMasterParamFileName, this.localizedNameID);
    }

    protected void init()
    {
      this.localizedNameID = this.GetType().GenerateLocalizedID(this.iname, "NAME");
    }

    public void Deserialize(string language, JSON_UnitParam json)
    {
      this.Deserialize(json);
      if (json.tag != null)
        this.tag = json.tag;
      this.localizeFields(language);
    }

    public void CacheReferences(MasterParam master)
    {
      if (this.jobsets == null || this.mJobSetCache != null)
        return;
      this.mJobSetCache = new JobSetParam[this.jobsets.Length];
      for (int index = 0; index < this.jobsets.Length; ++index)
        this.mJobSetCache[index] = master.GetJobSetParam((string) this.jobsets[index]);
    }

    public JobSetParam GetJobSetFast(int index)
    {
      if (this.mJobSetCache != null && 0 <= index && index < this.mJobSetCache.Length)
        return this.mJobSetCache[index];
      return (JobSetParam) null;
    }

    public string SexPrefix
    {
      get
      {
        return this.sex.ToPrefix();
      }
    }

    public bool Deserialize(JSON_UnitParam json)
    {
      if (json == null)
        return false;
      this.no = json.no;
      this.iname = json.iname;
      this.name = json.name;
      this.ai = (OString) json.ai;
      this.model = json.mdl;
      this.grow = (OString) json.grow;
      this.piece = (OString) json.piece;
      this.birth = (OString) json.birth;
      this.skill = (OString) json.skill;
      this.ability = (OString) json.ability;
      this.ma_quest = (OString) json.ma_quest;
      this.sw = (OInt) Math.Max(json.sw, 1);
      this.sh = (OInt) Math.Max(json.sh, 1);
      this.sex = (ESex) json.sex;
      this.rare = (OInt) json.rare;
      this.raremax = (OInt) json.raremax;
      this.type = (EUnitType) json.type;
      this.element = (EElement) json.elem;
      this.hero = (OInt) json.hero;
      this.search = (OInt) json.search;
      this.stopped = (OBool) (json.stop != 0);
      this.summon = json.notsmn == 0;
      if (!string.IsNullOrEmpty(json.available_at))
      {
        try
        {
          this.available_at = DateTime.Parse(json.available_at);
        }
        catch
        {
          this.available_at = DateTime.MaxValue;
        }
      }
      this.height = (OInt) json.height;
      this.weight = (OInt) json.weight;
      this.jobsets = (OString[]) null;
      this.mJobSetCache = (JobSetParam[]) null;
      this.tags = (string[]) null;
      this.jobtype = (JobTypes) json.jt;
      this.role = (RoleTypes) json.role;
      this.mov = (OInt) json.mov;
      this.jmp = (OInt) json.jmp;
      this.inimp = (OInt) json.inimp;
      this.ma_rarity = (OInt) json.ma_rarity;
      this.ma_lv = (OInt) json.ma_lv;
      if (json.skins != null && json.skins.Length >= 1)
      {
        this.skins = new OString[json.skins.Length];
        for (int index = 0; index < json.skins.Length; ++index)
          this.skins[index] = (OString) json.skins[index];
      }
      this.djob = (OString) json.djob;
      this.dbuki = (OString) json.dbuki;
      this.default_skill = (OString) json.dskl;
      this.default_abilities = (OString[]) null;
      if (json.dabi != null && json.dabi.Length > 0)
      {
        this.default_abilities = new OString[json.dabi.Length];
        for (int index = 0; index < json.dabi.Length; ++index)
          this.default_abilities[index] = (OString) json.dabi[index];
      }
      if (this.type == EUnitType.EventUnit)
        return true;
      if (json.jobsets != null)
      {
        this.jobsets = new OString[json.jobsets.Length];
        for (int index = 0; index < this.jobsets.Length; ++index)
          this.jobsets[index] = (OString) json.jobsets[index];
      }
      if (json.tag != null)
        this.tags = json.tag.Split(',');
      if (this.ini_status == null)
        this.ini_status = new BaseStatus();
      this.ini_status.param.hp = (OInt) json.hp;
      this.ini_status.param.mp = (OShort) json.mp;
      this.ini_status.param.atk = (OShort) json.atk;
      this.ini_status.param.def = (OShort) json.def;
      this.ini_status.param.mag = (OShort) json.mag;
      this.ini_status.param.mnd = (OShort) json.mnd;
      this.ini_status.param.dex = (OShort) json.dex;
      this.ini_status.param.spd = (OShort) json.spd;
      this.ini_status.param.cri = (OShort) json.cri;
      this.ini_status.param.luk = (OShort) json.luk;
      this.ini_status.enchant_resist.poison = (OInt) json.rpo;
      this.ini_status.enchant_resist.paralyse = (OInt) json.rpa;
      this.ini_status.enchant_resist.stun = (OInt) json.rst;
      this.ini_status.enchant_resist.sleep = (OInt) json.rsl;
      this.ini_status.enchant_resist.charm = (OInt) json.rch;
      this.ini_status.enchant_resist.stone = (OInt) json.rsn;
      this.ini_status.enchant_resist.blind = (OInt) json.rbl;
      this.ini_status.enchant_resist.notskl = (OInt) json.rns;
      this.ini_status.enchant_resist.notmov = (OInt) json.rnm;
      this.ini_status.enchant_resist.notatk = (OInt) json.rna;
      this.ini_status.enchant_resist.zombie = (OInt) json.rzo;
      this.ini_status.enchant_resist.death = (OInt) json.rde;
      this.ini_status.enchant_resist.knockback = (OInt) json.rkn;
      this.ini_status.enchant_resist.resist_buff = (OInt) 0;
      this.ini_status.enchant_resist.resist_debuff = (OInt) json.rdf;
      this.ini_status.enchant_resist.berserk = (OInt) json.rbe;
      this.ini_status.enchant_resist.stop = (OInt) json.rcs;
      this.ini_status.enchant_resist.fast = (OInt) json.rcu;
      this.ini_status.enchant_resist.slow = (OInt) json.rcd;
      this.ini_status.enchant_resist.donsoku = (OInt) json.rdo;
      this.ini_status.enchant_resist.rage = (OInt) json.rra;
      if (this.max_status == null)
        this.max_status = new BaseStatus();
      this.max_status.param.hp = (OInt) json.mhp;
      this.max_status.param.mp = (OShort) json.mmp;
      this.max_status.param.atk = (OShort) json.matk;
      this.max_status.param.def = (OShort) json.mdef;
      this.max_status.param.mag = (OShort) json.mmag;
      this.max_status.param.mnd = (OShort) json.mmnd;
      this.max_status.param.dex = (OShort) json.mdex;
      this.max_status.param.spd = (OShort) json.mspd;
      this.max_status.param.cri = (OShort) json.mcri;
      this.max_status.param.luk = (OShort) json.mluk;
      this.max_status.enchant_resist.poison = (OInt) json.mrpo;
      this.max_status.enchant_resist.paralyse = (OInt) json.mrpa;
      this.max_status.enchant_resist.stun = (OInt) json.mrst;
      this.max_status.enchant_resist.sleep = (OInt) json.mrsl;
      this.max_status.enchant_resist.charm = (OInt) json.mrch;
      this.max_status.enchant_resist.stone = (OInt) json.mrsn;
      this.max_status.enchant_resist.blind = (OInt) json.mrbl;
      this.max_status.enchant_resist.notskl = (OInt) json.mrns;
      this.max_status.enchant_resist.notmov = (OInt) json.mrnm;
      this.max_status.enchant_resist.notatk = (OInt) json.mrna;
      this.max_status.enchant_resist.zombie = (OInt) json.mrzo;
      this.max_status.enchant_resist.death = (OInt) json.mrde;
      this.max_status.enchant_resist.knockback = (OInt) json.mrkn;
      this.max_status.enchant_resist.resist_buff = (OInt) 0;
      this.max_status.enchant_resist.resist_debuff = (OInt) json.mrdf;
      this.max_status.enchant_resist.berserk = (OInt) json.mrbe;
      this.max_status.enchant_resist.stop = (OInt) json.mrcs;
      this.max_status.enchant_resist.fast = (OInt) json.mrcu;
      this.max_status.enchant_resist.slow = (OInt) json.mrcd;
      this.max_status.enchant_resist.donsoku = (OInt) json.mrdo;
      this.max_status.enchant_resist.rage = (OInt) json.mrra;
      this.leader_skills[0] = json.ls1;
      this.leader_skills[1] = json.ls2;
      this.leader_skills[2] = json.ls3;
      this.leader_skills[3] = json.ls4;
      this.leader_skills[4] = json.ls5;
      this.leader_skills[5] = json.ls6;
      this.recipes[0] = json.recipe1;
      this.recipes[1] = json.recipe2;
      this.recipes[2] = json.recipe3;
      this.recipes[3] = json.recipe4;
      this.recipes[4] = json.recipe5;
      this.recipes[5] = json.recipe6;
      this.image = json.img;
      this.voice = json.vce;
      if (json.jidx != null)
      {
        this.job_option_index = new OString[json.jidx.Length];
        for (int index = 0; index < json.jidx.Length; ++index)
          this.job_option_index[index] = (OString) json.jidx[index];
      }
      if (json.jimgs != null)
      {
        this.job_images = new OString[json.jimgs.Length];
        for (int index = 0; index < json.jimgs.Length; ++index)
          this.job_images[index] = (OString) json.jimgs[index];
      }
      if (json.jvcs != null)
      {
        this.job_voices = new OString[json.jvcs.Length];
        for (int index = 0; index < json.jvcs.Length; ++index)
          this.job_voices[index] = (OString) json.jvcs[index];
      }
      this.is_throw = (OBool) (json.no_trw == 0);
      return true;
    }

    public RecipeParam GetRarityUpRecipe(int rarity)
    {
      if (this.recipes == null)
        return (RecipeParam) null;
      if (rarity < 0 || rarity >= RarityParam.MAX_RARITY || rarity >= (int) this.raremax)
        return (RecipeParam) null;
      string recipe = this.recipes[rarity];
      if (string.IsNullOrEmpty(recipe))
        return (RecipeParam) null;
      return MonoSingleton<GameManager>.GetInstanceDirect().MasterParam.GetRecipeParam(recipe);
    }

    public static int GetLeaderSkillLevel(int rarity, int awakeLv)
    {
      return 1;
    }

    public static void CalcUnitLevelStatus(UnitParam unit, int unitLv, ref BaseStatus status)
    {
      GrowParam growParam = MonoSingleton<GameManager>.GetInstanceDirect().GetGrowParam((string) unit.grow);
      if (growParam != null && growParam.curve != null)
        growParam.CalcLevelCurveStatus(unitLv, ref status, unit.ini_status, unit.max_status);
      else
        unit.ini_status.CopyTo(status);
    }

    public static void CalcUnitElementStatus(EElement element, ref BaseStatus status)
    {
      FixParam fixParam = MonoSingleton<GameManager>.GetInstanceDirect().MasterParam.FixParam;
      int elementResistUpRate = (int) fixParam.ElementResistUpRate;
      int elementResistDownRate = (int) fixParam.ElementResistDownRate;
      EElement weakElement = UnitParam.GetWeakElement(element);
      EElement resistElement = UnitParam.GetResistElement(element);
      ElementParam elementResist1;
      EElement index1;
      (elementResist1 = status.element_resist)[index1 = weakElement] = (OShort) ((int) elementResist1[index1] + elementResistDownRate);
      ElementParam elementResist2;
      EElement index2;
      (elementResist2 = status.element_resist)[index2 = resistElement] = (OShort) ((int) elementResist2[index2] + elementResistUpRate);
    }

    public static EElement GetWeakElement(EElement type)
    {
      return UnitParam.WeakElements[(int) type];
    }

    public static EElement GetResistElement(EElement type)
    {
      return UnitParam.ResistElements[(int) type];
    }

    public int GetUnlockNeedPieces()
    {
      RarityParam rarityParam = MonoSingleton<GameManager>.GetInstanceDirect().MasterParam.GetRarityParam((int) this.rare);
      if (rarityParam != null)
        return (int) rarityParam.UnitUnlockPieceNum;
      return 0;
    }

    public bool CheckEnableUnlock()
    {
      return !string.IsNullOrEmpty((string) this.piece) && this.summon && (MonoSingleton<GameManager>.Instance.Player.FindUnitDataByUniqueParam(this) == null && MonoSingleton<GameManager>.Instance.Player.GetItemAmount((string) this.piece) >= this.GetUnlockNeedPieces());
    }

    public bool CheckAvailable(DateTime now)
    {
      return !(now < this.available_at);
    }

    public string GetJobVoice(string jobName)
    {
      return this.GetJobOptions(jobName, false);
    }

    public string GetJobImage(string jobName)
    {
      return this.GetJobOptions(jobName, true);
    }

    private string GetJobOptions(string jobName, bool isImage)
    {
      OString[] ostringArray = !isImage ? this.job_voices : this.job_images;
      if (string.IsNullOrEmpty(jobName) || ostringArray == null || this.job_option_index == null)
        return string.Empty;
      int index = Array.FindIndex<OString>(this.job_option_index, (Predicate<OString>) (p => (string) p == jobName));
      if (index < 0 || index >= ostringArray.Length)
        return string.Empty;
      return (string) ostringArray[index];
    }

    public string GetJobId(int jobIndex)
    {
      JobSetParam jobSetFast = this.GetJobSetFast(jobIndex);
      if (jobSetFast != null)
        return jobSetFast.job;
      return string.Empty;
    }
  }
}

File: Database/genConvert/UnitParam.py
def UnitParam(json):
    this={}#UnitParamjson)
    #if(json==null)
    #returnfalse
    if 'no' in json:
        this['no'] = json['no']
    if 'iname' in json:
        this['iname'] = json['iname']
    if 'name' in json:
        this['name'] = json['name']
    if 'ai' in json:
        this['ai'] = json['ai']
    if 'mdl' in json:
        this['model'] = json['mdl']
    if 'grow' in json:
        this['grow'] = json['grow']
    if 'piece' in json:
        this['piece'] = json['piece']
    if 'birth' in json:
        this['birth'] = json['birth']
    if 'birth_id' in json:
        this['birthID'] = json['birth_id']
        this['birth'] = json['birth']
    if 'skill' in json:
        this['skill'] = json['skill']
    if 'ability' in json:
        this['ability'] = json['ability']
    if 'ma_quest' in json:
        this['ma_quest'] = json['ma_quest']
    if 'sw' in json:
        this['sw'] = Math.Max(json['sw'],1)
    if 'sh' in json:
        this['sh'] = Math.Max(json['sh'],1)
    if 'sex' in json:
        this['sex'] = ENUM['ESex'][json['sex']]
    if 'rare' in json:
        this['rare'] = json['rare']
    if 'raremax' in json:
        this['raremax'] = json['raremax']
    if 'type' in json:
        this['type'] = ENUM['EUnitType'][json['type']]
    if 'elem' in json:
        this['element'] = ENUM['EElement'][json['elem']]
    if 'hero' in json:
        this['hero'] = json['hero']
    if 'search' in json:
        this['search'] = json['search']
    if 'stop' in json:
        this['stopped'] = (json['stop']!=0)
    if 'notsmn' in json:
        this['summon'] = json['notsmn']==0
    #if(!string.IsNullOrEmpty(json.available_at))
        #try
            if 'available_at' in json:
                this['available_at'] = DateTime.Parse
        #catch
            #this.available_at=DateTime.MaxValue
    if 'height' in json:
        this['height'] = json['height']
    if 'weight' in json:
        this['weight'] = json['weight']
    if 'jt' in json:
        this['jobtype'] = ENUM['JobTypes'][json['jt']]
    if 'role' in json:
        this['role'] = ENUM['RoleTypes'][json['role']]
    if 'mov' in json:
        this['mov'] = json['mov']
    if 'jmp' in json:
        this['jmp'] = json['jmp']
    if 'inimp' in json:
        this['inimp'] = json['inimp']
    if 'ma_rarity' in json:
        this['ma_rarity'] = json['ma_rarity']
    if 'ma_lv' in json:
        this['ma_lv'] = json['ma_lv']
        if 'skins' in json:
            this['skins'] = json['skins']
    #if(UnitParam.NoJobStatus.IsExistParam(json))
        #this.no_job_status=newUnitParam.NoJobStatus()
        #this.no_job_status.SetParam(json)
    #if(this.type==EUnitType.EventUnit)
    if 'djob' in json:
        this['djob'] = json['djob']
    if 'dbuki' in json:
        this['dbuki'] = json['dbuki']
    if 'dskl' in json:
        this['default_skill'] = json['dskl']
        if 'dabi' in json:
            this['default_abilities'] = newOString[json['dabi'].Length]
        if 'dabi' in json:
            this['default_abilities[index]'] = json['dabi'][index]
    #returntrue
    #if(json.jobsets!=null)
        #for(intindex=0index<this.jobsets.Length++index)
        if 'jobsets' in json:
            this['jobsets'] = json['jobsets']
    #if(json.tag!=null)
    if 'tag' in json:
        this['tags'] = json['tag'].Split
    #if(this.ini_status==null)
    #this.ini_status=newUnitParam.Status()
    #this.ini_status.SetParamIni(json)
    #this.ini_status.SetEnchentParamIni(json)
    #if(this.max_status==null)
    #this.max_status=newUnitParam.Status()
    #this.max_status.SetParamMax(json)
    #this.max_status.SetEnchentParamMax(json)
    this['ini_status']
    this['ini_status']['param']
    if 'hp' in json:
        this['ini_status']['param']['hp'] = json['hp']
    this['ini_status']['param']
    if 'mp' in json:
        this['ini_status']['param']['mp'] = json['mp']
    this['ini_status']['param']
    if 'atk' in json:
        this['ini_status']['param']['atk'] = json['atk']
    this['ini_status']['param']
    if 'def' in json:
        this['ini_status']['param']['def'] = json['def']
    this['ini_status']['param']
    if 'mag' in json:
        this['ini_status']['param']['mag'] = json['mag']
    this['ini_status']['param']
    if 'mnd' in json:
        this['ini_status']['param']['mnd'] = json['mnd']
    this['ini_status']['param']
    if 'dex' in json:
        this['ini_status']['param']['dex'] = json['dex']
    this['ini_status']['param']
    if 'spd' in json:
        this['ini_status']['param']['spd'] = json['spd']
    this['ini_status']['param']
    if 'cri' in json:
        this['ini_status']['param']['cri'] = json['cri']
    this['ini_status']['param']
    if 'luk' in json:
        this['ini_status']['param']['luk'] = json['luk']
    this['ini_status']
    this['ini_status']['enchant_resist']
    if 'rpo' in json:
        this['ini_status']['enchant_resist']['poison'] = json['rpo']
    this['ini_status']['enchant_resist']
    if 'rpa' in json:
        this['ini_status']['enchant_resist']['paralyse'] = json['rpa']
    this['ini_status']['enchant_resist']
    if 'rst' in json:
        this['ini_status']['enchant_resist']['stun'] = json['rst']
    this['ini_status']['enchant_resist']
    if 'rsl' in json:
        this['ini_status']['enchant_resist']['sleep'] = json['rsl']
    this['ini_status']['enchant_resist']
    if 'rch' in json:
        this['ini_status']['enchant_resist']['charm'] = json['rch']
    this['ini_status']['enchant_resist']
    if 'rsn' in json:
        this['ini_status']['enchant_resist']['stone'] = json['rsn']
    this['ini_status']['enchant_resist']
    if 'rbl' in json:
        this['ini_status']['enchant_resist']['blind'] = json['rbl']
    this['ini_status']['enchant_resist']
    if 'rns' in json:
        this['ini_status']['enchant_resist']['notskl'] = json['rns']
    this['ini_status']['enchant_resist']
    if 'rnm' in json:
        this['ini_status']['enchant_resist']['notmov'] = json['rnm']
    this['ini_status']['enchant_resist']
    if 'rna' in json:
        this['ini_status']['enchant_resist']['notatk'] = json['rna']
    this['ini_status']['enchant_resist']
    if 'rzo' in json:
        this['ini_status']['enchant_resist']['zombie'] = json['rzo']
    this['ini_status']['enchant_resist']
    if 'rde' in json:
        this['ini_status']['enchant_resist']['death'] = json['rde']
    this['ini_status']['enchant_resist']
    if 'rkn' in json:
        this['ini_status']['enchant_resist']['knockback'] = json['rkn']
    this['ini_status']['enchant_resist']
    if 'rdf' in json:
        this['ini_status']['enchant_resist']['resist_debuff'] = json['rdf']
    this['ini_status']['enchant_resist']
    if 'rbe' in json:
        this['ini_status']['enchant_resist']['berserk'] = json['rbe']
    this['ini_status']['enchant_resist']
    if 'rcs' in json:
        this['ini_status']['enchant_resist']['stop'] = json['rcs']
    this['ini_status']['enchant_resist']
    if 'rcu' in json:
        this['ini_status']['enchant_resist']['fast'] = json['rcu']
    this['ini_status']['enchant_resist']
    if 'rcd' in json:
        this['ini_status']['enchant_resist']['slow'] = json['rcd']
    this['ini_status']['enchant_resist']
    if 'rdo' in json:
        this['ini_status']['enchant_resist']['donsoku'] = json['rdo']
    this['ini_status']['enchant_resist']
    if 'rra' in json:
        this['ini_status']['enchant_resist']['rage'] = json['rra']
    this['ini_status']['enchant_resist']
    if 'rsa' in json:
        this['ini_status']['enchant_resist']['single_attack'] = json['rsa']
    this['ini_status']['enchant_resist']
    if 'raa' in json:
        this['ini_status']['enchant_resist']['area_attack'] = json['raa']
    this['ini_status']['enchant_resist']
    if 'rdc' in json:
        this['ini_status']['enchant_resist']['dec_ct'] = json['rdc']
    this['ini_status']['enchant_resist']
    if 'ric' in json:
        this['ini_status']['enchant_resist']['inc_ct'] = json['ric']
    this['']
    this['max_status']
    this['max_status']['param']
    if 'mhp' in json:
        this['max_status']['param']['hp'] = json['mhp']
    this['max_status']['param']
    if 'mmp' in json:
        this['max_status']['param']['mp'] = json['mmp']
    this['max_status']['param']
    if 'matk' in json:
        this['max_status']['param']['atk'] = json['matk']
    this['max_status']['param']
    if 'mdef' in json:
        this['max_status']['param']['def'] = json['mdef']
    this['max_status']['param']
    if 'mmag' in json:
        this['max_status']['param']['mag'] = json['mmag']
    this['max_status']['param']
    if 'mmnd' in json:
        this['max_status']['param']['mnd'] = json['mmnd']
    this['max_status']['param']
    if 'mdex' in json:
        this['max_status']['param']['dex'] = json['mdex']
    this['max_status']['param']
    if 'mspd' in json:
        this['max_status']['param']['spd'] = json['mspd']
    this['max_status']['param']
    if 'mcri' in json:
        this['max_status']['param']['cri'] = json['mcri']
    this['max_status']['param']
    if 'mluk' in json:
        this['max_status']['param']['luk'] = json['mluk']
    this['max_status']
    this['max_status']['enchant_resist']
    if 'mrpo' in json:
        this['max_status']['enchant_resist']['poison'] = json['mrpo']
    this['max_status']['enchant_resist']
    if 'mrpa' in json:
        this['max_status']['enchant_resist']['paralyse'] = json['mrpa']
    this['max_status']['enchant_resist']
    if 'mrst' in json:
        this['max_status']['enchant_resist']['stun'] = json['mrst']
    this['max_status']['enchant_resist']
    if 'mrsl' in json:
        this['max_status']['enchant_resist']['sleep'] = json['mrsl']
    this['max_status']['enchant_resist']
    if 'mrch' in json:
        this['max_status']['enchant_resist']['charm'] = json['mrch']
    this['max_status']['enchant_resist']
    if 'mrsn' in json:
        this['max_status']['enchant_resist']['stone'] = json['mrsn']
    this['max_status']['enchant_resist']
    if 'mrbl' in json:
        this['max_status']['enchant_resist']['blind'] = json['mrbl']
    this['max_status']['enchant_resist']
    if 'mrns' in json:
        this['max_status']['enchant_resist']['notskl'] = json['mrns']
    this['max_status']['enchant_resist']
    if 'mrnm' in json:
        this['max_status']['enchant_resist']['notmov'] = json['mrnm']
    this['max_status']['enchant_resist']
    if 'mrna' in json:
        this['max_status']['enchant_resist']['notatk'] = json['mrna']
    this['max_status']['enchant_resist']
    if 'mrzo' in json:
        this['max_status']['enchant_resist']['zombie'] = json['mrzo']
    this['max_status']['enchant_resist']
    if 'mrde' in json:
        this['max_status']['enchant_resist']['death'] = json['mrde']
    this['max_status']['enchant_resist']
    if 'mrkn' in json:
        this['max_status']['enchant_resist']['knockback'] = json['mrkn']
    this['max_status']['enchant_resist']
    if 'mrdf' in json:
        this['max_status']['enchant_resist']['resist_debuff'] = json['mrdf']
    this['max_status']['enchant_resist']
    if 'mrbe' in json:
        this['max_status']['enchant_resist']['berserk'] = json['mrbe']
    this['max_status']['enchant_resist']
    if 'mrcs' in json:
        this['max_status']['enchant_resist']['stop'] = json['mrcs']
    this['max_status']['enchant_resist']
    if 'mrcu' in json:
        this['max_status']['enchant_resist']['fast'] = json['mrcu']
    this['max_status']['enchant_resist']
    if 'mrcd' in json:
        this['max_status']['enchant_resist']['slow'] = json['mrcd']
    this['max_status']['enchant_resist']
    if 'mrdo' in json:
        this['max_status']['enchant_resist']['donsoku'] = json['mrdo']
    this['max_status']['enchant_resist']
    if 'mrra' in json:
        this['max_status']['enchant_resist']['rage'] = json['mrra']
    this['max_status']['enchant_resist']
    if 'mrsa' in json:
        this['max_status']['enchant_resist']['single_attack'] = json['mrsa']
    this['max_status']['enchant_resist']
    if 'mraa' in json:
        this['max_status']['enchant_resist']['area_attack'] = json['mraa']
    this['max_status']['enchant_resist']
    if 'mrdc' in json:
        this['max_status']['enchant_resist']['dec_ct'] = json['mrdc']
    this['max_status']['enchant_resist']
    if 'mric' in json:
        this['max_status']['enchant_resist']['inc_ct'] = json['mric']
    if 'ls1' in json:
        this['leader_skills'][0] = json['ls1']
    if 'ls2' in json:
        this['leader_skills'][1] = json['ls2']
    if 'ls3' in json:
        this['leader_skills'][2] = json['ls3']
    if 'ls4' in json:
        this['leader_skills'][3] = json['ls4']
    if 'ls5' in json:
        this['leader_skills'][4] = json['ls5']
    if 'ls6' in json:
        this['leader_skills'][5] = json['ls6']
    if 'recipe1' in json:
        this['recipes'][0] = json['recipe1']
    if 'recipe2' in json:
        this['recipes'][1] = json['recipe2']
    if 'recipe3' in json:
        this['recipes'][2] = json['recipe3']
    if 'recipe4' in json:
        this['recipes'][3] = json['recipe4']
    if 'recipe5' in json:
        this['recipes'][4] = json['recipe5']
    if 'recipe6' in json:
        this['recipes'][5] = json['recipe6']
    if 'img' in json:
        this['image'] = json['img']
    if 'vce' in json:
        this['voice'] = json['vce']
    #this.flag.Set(2,json.no_trw==0)
    #this.flag.Set(3,json.no_kb==0)
    if 'unlck_t' in json:
        this['unlock_time'] = json['unlck_t']
        if 'jidx' in json:
            this['job_option_index'] = newOString[json['jidx'].Length]
        if 'jidx' in json:
            this['job_option_index[index]'] = json['jidx'][index]
        if 'jimgs' in json:
            this['job_images'] = newOString[json['jimgs'].Length]
        if 'jimgs' in json:
            this['job_images[index]'] = json['jimgs'][index]
        if 'jvcs' in json:
            this['job_voices'] = newOString[json['jvcs'].Length]
        if 'jvcs' in json:
            this['job_voices[index]'] = json['jvcs'][index]
    if 'no_trw' in json:
        this['is_throw'] = (json['no_trw']==0)
    if 'no_kb' in json:
        this['is_knock_back'] = (json['no_kb']==0)
    #returntrue
return this
